crates/autopilot/src/domain/competition/winner_selection/combinatorial.rs

```diff
@@ -350,3 +350,798 @@ type ScoreByDirection = HashMap<DirectedTokenPair, Score>;
+            "tokens": [
+                ["Token A", address(0)],
+                ["Token B", address(1)],
+                ["Token C", address(2)],
+                ["Token D", address(3)]
+            ],
+            "auction": {
+                "orders": {
+                    "Order 1": ["sell", "Token A", amount(1_000), "Token B", amount(1_000)],
+                    "Order 2": ["sell", "Token C", amount(1_000), "Token D", amount(1_000)],
+                }
+            },
+            "solutions": {
+                // unfair batch
+                "Solution 1": {
+                    "solver": "Solver 1",
+                    "trades": {
+                        "Order 1": [amount(1_000), amount(1_100)],
+                        "Order 2": [amount(1_000), amount(1_100)]
+                    },
+                    "score": score(200),
+                },
+                // filtering batch
+                "Solution 2": {
+                    "solver": "Solver 2",
+                    "trades": {
+                        "Order 1": [amount(1_000), amount(1_150)],
+                    },
+                    "score": score(150),
+                }
+            },
+            "expected_fair_solutions": ["Solution 2"],
+            "expected_winners": ["Solution 2"],
+            "expected_reference_scores": {
+                "Solver 2": "0",
+            },
+        });
+        TestCase::from_json(case).validate();
+    }
+
+    #[test]
+    // Multiple trades on the same (directed) token pair are aggregated for
+    // filtering
+    fn aggregation_on_token_pair() {
+        let case = json!({
+            "tokens": [
+                ["Token A", address(0)],
+                ["Token B", address(1)],
+            ],
+            "auction": {
+                "orders": {
+                    "Order 1": ["sell", "Token A", amount(1_000), "Token B", amount(1_000)],
+                    "Order 2": ["sell", "Token A", amount(1_000), "Token B", amount(1_000)],
+                }
+            },
+            "solutions": {
+                // batch with aggregation
+                "Solution 1": {
+                    "solver": "Solver 1",
+                    "trades": {
+                        "Order 1": [amount(1_000), amount(1_100)],
+                        "Order 2": [amount(1_000), amount(1_100)]
+                    },
+                    "score": score(200),
+                },
+                // incompatible batch
+                "Solution 2": {
+                    "solver": "Solver 2",
+                    "trades": {
+                        "Order 1": [amount(1_000), amount(1_150)],
+                    },
+                    "score": score(150),
+                }
+            },
+            "expected_fair_solutions": ["Solution 1", "Solution 2"],
+            "expected_winners": ["Solution 1"],
+            "expected_reference_scores": {
+                "Solver 1": "150",
+                "Solver 2": "200",
```

Comment:
Also here we should only have a reference score for solver 1. The test assertions are not tight enough and there is possibly a bug in the reference score computation logic.